src/cache/cache.go

```diff
@@ -0,0 +1,121 @@
+package cache
+
+import (
+	"os"
+	"sync"
+	"../heap"
+	"../markov"
+)
+
+/********************************
+Cache supports the following external API to users
+
+c.Init(cacheSize int, cacheType CacheType)
+    Initializes a cache with eviction policy and prefetch defined by cache type
+c.Report() (hits, misses)
+    Get a report of the hits and misses  TODO: Do we want a version number or
+    timestamp mechanism of any form here?
+c.Fetch(name string) (*os.File, error)
+
+*********************************/
+type Cache struct {
+	mu          sync.Mutex          // Lock to protect shared access to cache
+	misses		int
+	hits		int
+	cache		map[string]*os.File
+	heap		heap.MinHeap
+	timestamp	int64 // for controlling LRU heap
+	cacheSize	int
+	chain 		*markov.MarkovChain
+	cacheType	CacheType
+}
+
+
+func (c *Cache) Init(cacheSize int, cacheType CacheType) {
+	c.cacheType = cacheType
+	c.misses = 0
+	c.hits = 0
+    c.cacheSize = cacheSize
+	c.cache = make(map[string]*os.File)
+	c.timestamp = 0
+	
+	if cacheType == LRU || cacheType == MarkovEviction {
+		// only LRU caches should use heap
+		c.heap.Init()
+	}
+	if cacheType != LRU {
+		// all other caches need a MarkovChain
+		c.chain = markov.MakeMarkovChain()
+	}
+}
+
+func (c *Cache) Fetch(name string) (*os.File, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	file, ok := c.cache[name]
+	var err error
```

Comment:
tiny nit: If `error` is a type defined by us, can we capitalize?

Reply:
Not defined by us, default golang interface: https://blog.golang.org/error-handling-and-go